applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py

```diff
@@ -0,0 +1,157 @@
+
+        self.signed_distances = None
+        self.directions = None
+
+        self.value = None
+
+        self.gradient = {}
+
+        self.infeasible_side = self.response_settings["infeasible_side"].GetBool()
+        self.exponent = 2
+
+    @classmethod
+    def GetDefaultSettings(cls):
+        this_defaults = KM.Parameters("""{
+            "response_type"         : "plane_packaging",
+            "model_part_name"       : "UNKNOWN_NAME",
+            "domain_size"           : 3,
+            "model_import_settings" : {
+                "input_type"        : "use_input_model_part",
+                "input_filename"    : "UNKNOWN_NAME"
+            },
+            "infeasible_side"       : false
+        }""")
+        return this_defaults
+
+    def Initialize(self):
+        if self.model_part_needs_to_be_imported:
+            file_name = self.response_settings["model_import_settings"]["input_filename"].GetString()
+            model_part_io = KratosMultiphysics.ModelPartIO(file_name)
+            model_part_io.ReadModelPart(self.model_part)
+
+    def InitializeSolutionStep(self):
+        self.value = None
+        self.signed_distances = None
+        self.directions = None
+        self.gradient = {}
+
+    def CalculateValue(self):
+        Logger.PrintInfo("\n> Starting primal analysis for response", self.identifier)
+
+        startTime = timer.time()
+
+        if not self.directions or not self.signed_distances:
+            self._CalculateProjectedDistances()
+
+        value = 0.0
+        for i in range(len(self.signed_distances)):
+            value += self._CalculateValueForNode(self.signed_distances[i], self.directions[i*3:i*3+3])
+
+        self.value = value
+
+        Logger.PrintInfo("> Time needed for calculating the response value = ", round(timer.time() - startTime,2), "s")
+
+    def CalculateGradient(self):
+        Logger.PrintInfo("\n> Starting gradient calculation for response", self.identifier)
+
+        startTime = timer.time()
+
+        if not self.directions or not self.signed_distances:
+            self._CalculateProjectedDistances()
+
+        for i, node in enumerate(self.model_part.Nodes):
+            gradient = self._CalculateGradientForNode(self.signed_distances[i], self.directions[i*3:i*3+3])
+            self.gradient[node.Id] = gradient
+
+        Logger.PrintInfo("> Time needed for calculating gradients = ", round(timer.time() - startTime,2), "s")
+
+    def GetValue(self):
+        return self.value
+
+    def GetShapeGradient(self):
+        if not self.gradient:
+            raise RuntimeError("Gradient was not calculated")
+        return self.gradient
+
+    def _CalculateProjectedDistances(self):
+        raise NotImplementedError("_CalculateProjectedDistances needs to be implemented by the derived class!")
+
+    def _CalculateValueForNode(self, signed_distance, direction):
```

Comment:
What about the following names for this function and the one below:

`_CalculateNodalValue`
`_CalculateNodalGradient`

Your call.